### PR TITLE
Introduce 'SECURE' vulnerability variant handled both on API and UI

### LINKED ISSUE #258: Adding a way to include some indicator for FalsePositives
**Is your feature request related to a problem? Please describe.**
Currently we can add the vulnerabilities in the VulnerableApp but there is no way to have some false positives "Secure" implementations such that we can better judge the scanners.

**Additional context**
we have a level which is called "secure" but we might need something like "secure1" , "secure2" etc. so that those can be secure implementations and "/scanner" endpoint will return those and then in Owasp ZAP evaluation script we can include it.

Owasp ZAP Integration with VulnerableApp: https://github.com/zapbot/zap-mgmt-scripts/tree/master/vulnerableApp

### PR DESCRIPTION
**Pull request by nowakkamil**
This pull request closes #258.
This change is in line with: https://github.com/SasanLabs/VulnerableApp/issues/258#issuecomment-826163238.

### CODE REVIEW

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/beans/LevelResponseBean.java:13]: can you please run `gradle spotlessApply` ?

File: src/main/java/org/sasanlabs/internal/utility/SecureConstants.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/internal/utility/SecureConstants.java:6]: As discussed, we don't need SecureConstants as we are relying on Level and Variant(Type of Secure/Unsecure)

File: src/main/java/org/sasanlabs/internal/utility/VariantConstants.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/internal/utility/VariantConstants.java:3]: can we create an Enum ? Also please add Author tag and description of Enum.

File: src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java:32]: Please go ahead with Enum.

File: src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java:209]: make value = LevelConstants.LEVEL_*

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:77]: we can rename it to `_getSvgElementForVariant` or some other name which depicts that it is returning the javascript element. "_" is to depict that it is a private utility.

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:80]: can we go with if else ?

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:73]: start the name with _ to depict that it is a private utility method.

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:128]: please add a description of method and also variable id.

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:133]: as id is an integer so please add a proper condition like `id === 0`

**Review by preetkaran20 — CHANGES_REQUESTED** [review]
Please add some screenshots/screen video depicting the change so that i can know what exactly this pr has changed in UI

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/beans/LevelResponseBean.java:120]: As we are not using SecureConstants so this change might not be required. please check.

File: src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java:32]: can we name it as SecurityVariant or SecureVariant ?

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/beans/LevelResponseBean.java]: Resolved in 422b8d3

File: src/main/resources/static/vulnerableApp.js
Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js]: Resolved in 422b8d3

File: src/main/resources/static/vulnerableApp.js
Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js]: Resolved in 422b8d3

File: src/main/resources/static/vulnerableApp.js
Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js]: Resolved in 422b8d3

File: src/main/resources/static/vulnerableApp.js
Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js:137]: Resolved in 422b8d3

File: src/main/resources/static/vulnerableApp.js
Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js]: Resolved in 422b8d3

File: src/main/java/org/sasanlabs/internal/utility/VariantConstants.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/internal/utility/VariantConstants.java:3]: Unfortunately, it's not possible due to Java limitations. I tried the ```Enum``` approach initially, but the annotation requires providing compile-time constant value, so I had to abandon this idea. The current implementation is similar to what we already have for ```LevelConstants```.

File: src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java:209]: Doing that results in a run-time exception thrown by having the same value in the controller scope. This behaviour prevents ambiguity. Introducing duplicates would mean that identical endpoint would have to be handled via different methods. Additonally, this is connected to https://github.com/SasanLabs/VulnerableApp/pull/278#discussion_r623483099

File: src/main/java/org/sasanlabs/internal/utility/SecureConstants.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/internal/utility/SecureConstants.java:4]: Please take a look at https://github.com/SasanLabs/VulnerableApp/pull/278#discussion_r623483099 and https://github.com/SasanLabs/VulnerableApp/pull/278#discussion_r623490100

File: src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java:31]: I've got a mild preference for the current approach as I believe this _security_ part might be inferred by looking at sibling and parent properties. However, if you feel like your approach could avoid confusion in the future, I'm happy to make an update. Please let me know your final decision.

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/beans/LevelResponseBean.java]: Please take a look at https://github.com/SasanLabs/VulnerableApp/pull/278#discussion_r623483099 and https://github.com/SasanLabs/VulnerableApp/pull/278#discussion_r623490100

**Comment by nowakkamil** [comment]
The screenshot as per @preetkaran20 [request](https://github.com/SasanLabs/VulnerableApp/pull/278#pullrequestreview-645315758)
image

File: src/main/java/org/sasanlabs/internal/utility/VariantConstants.java
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/internal/utility/VariantConstants.java:3]: Enums are allowed in annotations and they are static constants only. Look at following line https://github.com/SasanLabs/VulnerableApp/blob/ed98eb0923893ea822761cca3e7d0969b0eb340d/src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java#L48 for more information.

File: src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/internal/utility/annotations/VulnerableAppRequestMapping.java:31]: then lets go with variant only but add a javadoc explaining the use of this field

File: src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java:209]: no no wait, * was the next level number i.e. `Level_7`. please try this and it will work.

File: src/main/java/org/sasanlabs/internal/utility/SecureConstants.java
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/internal/utility/SecureConstants.java:4]: please have a look at https://github.com/SasanLabs/VulnerableApp/pull/278/files#r623531869

**Review by preetkaran20 — CHANGES_REQUESTED** [review]
added few comments. Also as there are lots of UI changes, please do some testing such that it doesn't break any existing flows.

File: src/main/java/org/sasanlabs/internal/utility/VariantConstants.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/internal/utility/VariantConstants.java]: I see your point. I tried with different Enum (containing String), but now the simplified approach works. Resolved in 7eca817

File: src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java:209]: It will, but wasn't this endpoint supposed to be marked as _secure_?

File: src/main/java/org/sasanlabs/internal/utility/SecureConstants.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/internal/utility/SecureConstants.java]: Resolved in 3211f26

File: src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/service/vulnerability/xss/reflected/UrlParamBasedImgTagAttrInjection.java]: As recently discussed, a single level will be unique in the controller scope, so I have changed it per your suggesion. Resolved in 3211f26

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/beans/LevelResponseBean.java]: Resolved in 3211f26

File: src/main/java/org/sasanlabs/beans/LevelResponseBean.java
Comment (nowakkamil) [reply src/main/java/org/sasanlabs/beans/LevelResponseBean.java]: Resolved in 3211f26

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [inline src/main/resources/static/vulnerableApp.js:73]: name can be `_isSecureVariant`

**Review by preetkaran20 — APPROVED** [review]

File: src/main/resources/static/vulnerableApp.js
Comment (preetkaran20) [reply src/main/resources/static/vulnerableApp.js:73]: @nowakamil

Comment (nowakkamil) [reply src/main/resources/static/vulnerableApp.js]: Resolved in 0c0d74c

**Comment by nowakkamil** [comment]
@preetkaran20, please review the final version and possibly merge if everything seems sorted out from your perspective. I've added a tooltip as previously discussed and tested the UI. As a result, no issues have been found. The following screenshot depicts the latest version
image

**Comment by preetkaran20** [comment]
> @preetkaran20, please review the final version and possibly merge if everything seems sorted out from your perspective. I've added a tooltip as previously discussed and tested the UI. As a result, no issues have been found. The following screenshot depicts the latest version
> image

Can we have image color as green and hover over level as black and tooltip text as white and background color as black. Also tooltip text looks a lot bigger, can we have it as "Secure Implementation"

**Comment by nowakkamil** [comment]
I've updated the PR by making the tooltip font a little bit smaller and changing its text. Below are two tooltip variants:
1. Unsecure
image

2. Secure
image


Regarding making the secure level text black, I think it is more congruent to have the vector and text in the same green colour. It also looks better, in my opinion.

**Comment by preetkaran20** [comment]
> I've updated the PR by making the tooltip font a little bit smaller and changing its text. Below are two tooltip variants:
> 
> 1. Unsecure
>    image
> 2. Secure
>    image
> 
> Regarding making the secure level text black, I think it is more congruent to have the vector and text in the same green colour. It also looks better, in my opinion.

Yeah makes sense.

**Comment by preetkaran20** [comment]
Before merging, i will take the update from your fork for little testing. In the mean time please go ahead with the VulnerableApp-Nginx changes as discussed earlier over the call.

**Comment by preetkaran20** [comment]
@nowakkamil I have verified the PR and it looks good. Merging it.

Thanks a lot for the contribution. Now we are left with the next task which is making the same changes in VulnerableApp-Nginx.

thanks and regards,
Karan

**Pull request merged**
